File: backend/src/main/java/com/fooddelivery/service/SearchService.java
package com.fooddelivery.service;

import com.fooddelivery.dto.FoodItemResponse;
import com.fooddelivery.dto.RestaurantResponse;
import com.fooddelivery.entity.Category;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;

/**
 * Service interface for search and recommendation operations.
 */
public interface SearchService {

    // ---- Search ----

    Page<RestaurantResponse> searchRestaurants(String keyword, Double minRating,
                                                String sortBy, Pageable pageable);

    Page<FoodItemResponse> searchFoods(String keyword, Category category, Boolean veg,
                                       BigDecimal minPrice, BigDecimal maxPrice, Pageable pageable);

    // ---- Recommendations ----

    List<RestaurantResponse> getTopRatedRestaurants(int limit);

    List<FoodItemResponse> getTrendingFoods(int limit);
}

File: frontend/src/services/searchService.js
/**
 * Search & Recommendations API service — maps to SearchController.
 */
import api from '../api/axios';

const searchService = {
  /**
   * Search restaurants by keyword, minRating, with pagination.
   */
  searchRestaurants: (params = {}) =>
    api.get('/api/search/restaurants', { params }).then((r) => r.data),

  /**
   * Search food items with filters: keyword, category, veg, minPrice, maxPrice.
   */
  searchFoods: (params = {}) =>
    api.get('/api/search/foods', { params }).then((r) => r.data),

  /**
   * Get top-rated restaurants.
   */
  getTopRestaurants: (limit = 10) =>
    api.get('/api/recommendations/top-restaurants', { params: { limit } }).then((r) => r.data),

  /**
   * Get trending (most ordered) food items.
   */
  getTrendingFoods: (limit = 10) =>
    api.get('/api/recommendations/trending-foods', { params: { limit } }).then((r) => r.data),
};

export default searchService;
